rust/src/fvm/engine.rs

```diff
@@ -180,10 +184,302 @@ mod v3 {
+                        gas_used: ret
+                            .msg_receipt
+                            .gas_used
+                            .try_into()
+                            .context("negative gas used")?,
+                        events_root: ret.msg_receipt.events_root,
+                    },
+                    penalty: TokenAmount::from_atto(ret.penalty.atto().clone()),
+                    miner_tip: TokenAmount::from_atto(ret.miner_tip.atto().clone()),
+                    base_fee_burn: TokenAmount::from_atto(ret.base_fee_burn.atto().clone()),
+                    over_estimation_burn: TokenAmount::from_atto(
+                        ret.over_estimation_burn.atto().clone(),
+                    ),
+                    refund: TokenAmount::from_atto(ret.refund.atto().clone()),
+                    gas_refund: ret.gas_refund.try_into().context("negative gas refund")?,
+                    gas_burned: ret.gas_burned.try_into().context("negative gas burned")?,
+                    failure_info: ret.failure_info.map(|failure| match failure {
+                        ApplyFailure3h1::MessageBacktrace(bt) => {
+                            ApplyFailure::MessageBacktrace(Backtrace {
+                                frames: bt
+                                    .frames
+                                    .into_iter()
+                                    .map(|f| Frame {
+                                        source: f.source,
+                                        method: f.method,
+                                        code: ExitCode::new(f.code.value()),
+                                        message: f.message,
+                                    })
+                                    .collect(),
+                                cause: bt.cause.map(|cause| match cause {
+                                    Cause3h1::Syscall {
+                                        module,
+                                        function,
+                                        error,
+                                        message,
+                                    } => Cause::Syscall {
+                                        module,
+                                        function,
+                                        error: ErrorNumber::from_u32(error as u32)
+                                            .unwrap_or(ErrorNumber::AssertionFailed),
+                                        message,
+                                    },
+                                    Cause3h1::Fatal {
+                                        error_msg,
+                                        backtrace,
+                                    } => Cause::Fatal {
+                                        error_msg,
+                                        backtrace,
+                                    },
+                                }),
+                            })
```

Comment:
Do we have actual data structure changes here? If not, is not possible to transmute the structs?

Reply:
Transmuting without specified representation is always UB. Rust can,e.g., decide to re-order enum tags based on usage (it probably won't, but I don't want to play that game).